pkg/operatorstatus/monitor.go

```diff
@@ -0,0 +1,215 @@
+func initInstance() error {
+	// Get the watch namespace.
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		return fmt.Errorf("failed to get watch namespace: %v", err)
+	}
+
+	// Create a new writer.
+	coWriter, err := newWriter()
+	if err != nil {
+		return fmt.Errorf("Error creating ClusterOpreator writer: %v", err)
+	}
+
+	// Initialize the instance singleton once.
+	once.Do(func() {
+		// init the instance singleton.
+		instance = &statusMonitor{
+			coWriter:     coWriter,
+			namespace:    namespace,
+			eventTracker: eventTracker{0, 0, 1000, 10, sync.Mutex{}},
+			eventCh:      make(chan error, 32),
+		}
+	})
+
+	return nil
+}
+
+// StartReporting initializes the instance singleton and starts reporting
+// ClusterOperator status.
+func StartReporting() error {
+	err := initInstance()
+	if err != nil {
+		return err
+	}
+
+	err = instance.reportProgressing()
+	if err != nil {
+		log.Errorf("[status] Error updating status: %v", err)
+	}
+
+	go instance.eventChannelReceiver()
+	go instance.monitorEvents()
+
+	return nil
+}
+
+// eventChannelReceiver will listen on the event channel and update the
+// number of syncs and failed phase transition events appropriately.
+func (s *statusMonitor) eventChannelReceiver() {
+	log.Info("[status] Starting event receiver")
+	for {
+		select {
+		case err := <-s.eventCh:
+			if err == nil {
+				s.eventTracker.incrementSyncs()
+			} else {
+				s.eventTracker.incrementFailedTransitions()
+			}
+		}
+	}
+}
+
+// SendEventMessage is used to send events to the eventCh. If the channel is
+// busy, the event will be dropped to prevent the controller from stalling.
+func SendEventMessage(err error) {
+	// If the coAPI is not available do not attempt to send messages to the
+	// sync channel
+	if instance == nil {
+		return
+	}
+
+	// A missing sync status is better than stalling the controller
+	select {
+	case instance.eventCh <- err:
+		break
+	default:
+		log.Warning("[status] Event channel is busy, not reporting event")
+	}
+}
+
+// monitorEvents will update the ClusterOpreator to reflect that Marketplace
+// is available or degraded based on the events reported.
+func (s *statusMonitor) monitorEvents() {
+	// reportInterval represents the frequency at which marketplace
+	// will attempt to update the ClusterOperator.
+	reportInterval := 1 * time.Minute
+	for {
+		select {
+		case <-time.After(reportInterval):
+			var statusErr error
+			defer func() {
+				if statusErr != nil {
+					log.Errorf("[status] Error updating status: %v", statusErr)
+				}
+			}()
+			syncs, failures := s.eventTracker.getEvents()
+
+			// Wait for the minimum number of sync events.
+			if syncs < minimumSyncs {
+				break
+			}
+
+			// Check if the ratio is appropriate.
+			if syncs > 0 && syncs >= minimumSyncs {
+				if float64(failures)/float64(syncs) < degradedRatio {
+					statusErr = s.reportAvailable()
+				} else {
+					statusErr = s.reportDegraded()
+				}
+				reportInterval = 5 * time.Minute
```

Comment:
Constant again.

Reply:
Will update based on feedback from [this thread](https://github.com/operator-framework/operator-marketplace/pull/212#discussion_r301708497)